coreneuron/io/core2nrn_data_return.cpp

```diff
@@ -145,7 +243,272 @@ void core2nrn_data_return() {
+        if (ps.thvar_index_ >= 0) {
+            int index_v = pinv_nt ? pinv_nt[ps.thvar_index_] : ps.thvar_index_;
+            if (presyns_flag_true.erase(index_v)) {
+                psh.flag_ = 1;
+                if (presyns_flag_true.empty()) {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+static void core2nrn_tqueue_item(TQItem* q, NrnThread& nt) {
+    DiscreteEvent* d = (DiscreteEvent*) q->data_;
+    double td = q->t_;
+
+    // potentially several SelfEvent TQItem* associated with same weight index.
+    std::map<int, std::vector<TQItem*>> self_event_weight_map;
+
+    switch (d->type()) {
+        case NetConType: {
+            NetCon* nc = (NetCon*) d;
+            assert(nc >= nt.netcons && (nc < (nt.netcons + nt.n_netcon)));
+            size_t nc_index = nc - nt.netcons;
+            (*core2nrn_NetCon_event_)(nt.id, td, nc_index);
+            break;
+        }
+        case SelfEventType: {
+            SelfEvent* se = (SelfEvent*) d;
+            Point_process* pnt = se->target_;
+            assert(pnt->_tid == nt.id);
+            int tar_type = (int) pnt->_type;
+            int tar_index = pnt->_i_instance;
+            double flag = se->flag_;
+            TQItem** movable = (TQItem**) (se->movable_);
+            int is_movable = (movable && *movable == q) ? 1 : 0;
+            int weight_index = se->weight_index_;
+            // the weight_index is useless on the NEURON side so we need
+            // to convert that to NetCon index  and let the NEURON side
+            // figure out the weight_index. To figure out the netcon_index
+            // construct a {weight_index : [TQItem]} here for any
+            // weight_index >= 0, otherwise send it NEURON now.
+            if (weight_index >= 0) {
+                self_event_weight_map[weight_index].push_back(q);
+            } else {
+                (*core2nrn_SelfEvent_event_noweight_)(
+                    nt.id, td, tar_type, tar_index, flag, is_movable);
+            }
+            break;
         }
+        case PreSynType: {
+            PreSyn* ps = (PreSyn*) d;
+            printf("PreSynType %g\n", td);
+            break;
+        }
+        case NetParEventType: {
+            // nothing to transfer
+            break;
+        }
+        case PlayRecordEventType: {
+            // nothing to transfer
+            break;
+        }
+        default: {
+            // In particular, InputPreSyn does not appear in tqueue as it
+            // immediately fans out to NetCon.
+            assert(0);
+            break;
+        }
+    }
+
+    // For self events with weight, find the NetCon index and send that
+    // to NEURON.
+    if (!self_event_weight_map.empty()) {
+        for (int nc_index = 0; nc_index < nt.n_netcon; ++nc_index) {
+            NetCon& nc = nt.netcons[nc_index];
+            int weight_index = nc.u.weight_index_;
+            auto search = self_event_weight_map.find(weight_index);
```

Comment:
```suggestion
            const auto& search = self_event_weight_map.find(weight_index);
```

Reply:
👎 from me; IMO cleaner to use `auto` (or `auto const`) for things returned by value.

Reply:
In this specific case, agreed